File: apps/frontend/src/lib/api/public/carousel.ts
import apiClient from "../client";

export interface CarouselImage {
    _id: string;
    title?: string;
    subtitle?: string;
    image_url?: string;
    link_url?: string;
    display_order: number;
    active: boolean;
    created_at: string;
    updated_at: string;
}

export interface CarouselImagesListResponse {
    images: CarouselImage[];
    total: number;
    page: number;
    page_size: number;
}

/**
 * Get active carousel images for public display
 */
export async function getActiveCarouselImages(): Promise<CarouselImage[]> {
    try {
        const response = await apiClient.get<CarouselImagesListResponse>(
            "/api/v1/carousel/",
            {
                params: {
                    active: true,
                    page: 1,
                    page_size: 100,
                },
            }
        );
        return response.data.images;
    } catch (error) {
        console.error("Failed to fetch carousel images:", error);
        return [];
    }
}

File: apps/backend/app/schemas/carousel.py
from pydantic import BaseModel, HttpUrl, Field
from typing import Optional, List
from datetime import datetime


class CarouselImageCreate(BaseModel):
    """Schema for creating a new carousel image"""
    title: Optional[str] = None
    subtitle: Optional[str] = None
    link_url: Optional[HttpUrl] = None
    display_order: Optional[int] = Field(None, ge=0, description="Display order (0 or higher)")
    active: bool = True


class CarouselImageUpdate(BaseModel):
    """Schema for updating carousel image metadata"""
    title: Optional[str] = None
    subtitle: Optional[str] = None
    link_url: Optional[HttpUrl] = None
    display_order: Optional[int] = Field(None, ge=0)
    active: Optional[bool] = None


class CarouselImageResponse(BaseModel):
    """Schema for carousel image API response"""
    _id: str
    title: Optional[str] = None
    subtitle: Optional[str] = None
    image_url: Optional[str] = None
    link_url: Optional[str] = None
    display_order: int
    active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CarouselImagesListResponse(BaseModel):
    """Schema for paginated list of carousel images"""
    images: List[CarouselImageResponse]
    total: int
    page: int
    page_size: int


class UploadResponse(BaseModel):
    """Schema for image upload response"""
    url: str
    message: str


class ReorderRequest(BaseModel):
    """Schema for batch reordering carousel images"""
    image_orders: List[dict] = Field(..., description="List of {id: str, display_order: int}")
